### PR TITLE
Ouroboros wallet

### PR DESCRIPTION
**Pull request by i-Alex**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import com.horizen.utils.{BoxMerklePathInfo, BoxMerklePathInfoSerializer, ByteArrayWrapper, ListSerializer, Pair}
+import scorex.util.ScorexLogging
+import scorex.crypto.hash.Blake2b256
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Random, Success, Try}
+import java.util.{ArrayList => JArrayList}
+
+
+class ForgingBoxesMerklePathStorage(storage: Storage) extends SidechainTypes with ScorexLogging
```
Comment (oaleksandrov-ip) [inline sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: comment with version update logic description is required.

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala [hunk 2]
```diff
@@ -126,7 +130,9 @@ class SidechainWallet private[horizen] (seed: Array[Byte], walletBoxStorage: Sid
 
     walletBoxStorage.update(new ByteArrayWrapper(version), newBoxes.toList, boxIdsToRemove.toList).get
 
-    walletTransactionStorage.update(new ByteArrayWrapper(version), transactions)
+    walletTransactionStorage.update(new ByteArrayWrapper(version), transactions).get
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainWallet.scala]: I think, `.get` is used to transmit the exception if happen, but `require()` - not.
let's discuss

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala [hunk 3]
```diff
@@ -194,26 +201,62 @@ class SidechainWallet private[horizen] (seed: Array[Byte], walletBoxStorage: Sid
 
   override def walletSeed(): Array[Byte] = seed
 
-  def applyConsensusEpochInfo(epochInfo: ConsensusEpochInfo): SidechainWallet = this
+  def applyConsensusEpochInfo(epochInfo: ConsensusEpochInfo): SidechainWallet = {
+    val merkleTreeLeaves = epochInfo.forgersBoxIds.leaves().asScala.map(leaf => new ByteArrayWrapper(leaf))
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainWallet.scala]: it was a suggestion from Alberto during NodeViewHolder review.

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala [hunk 4]
```diff
@@ -194,26 +201,62 @@ class SidechainWallet private[horizen] (seed: Array[Byte], walletBoxStorage: Sid
...
+    forgingBoxesMerklePathStorage.update(epochInfo.epoch, boxMerklePathInfoSeq)
+    this
+  }
+
+  def getForgingBoxMerklePath(forgingBoxId: Array[Byte], requestedEpoch: ConsensusEpochNumber): Option[MerklePath] = {
+    // For given epoch N we should get data from the ending of the epoch N-2.
+    // genesis block is the single and the last block of epoch 1 - that is a special case:
+    // Data from epoch 1 is also valid for epoch 2, so for epoch N==2, we should get info from epoch 1.
+    val storedConsensusEpochNumber: ConsensusEpochNumber = requestedEpoch match {
+      case epoch if epoch <= 2 => ConsensusEpochNumber @@ 1
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainWallet.scala]: agree (but this looks funny)

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala [hunk 5]
```diff
@@ -194,26 +201,62 @@ class SidechainWallet private[horizen] (seed: Array[Byte], walletBoxStorage: Sid
...
+      case epoch => ConsensusEpochNumber @@ (epoch - 2)
+    }
+
+    forgingBoxesMerklePathStorage.getMerklePathsForEpoch(storedConsensusEpochNumber) match {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainWallet.scala]: agree

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala (hunk 1 above)
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: agree

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    val version = new Array[Byte](32)
+    lastVersionId match {
+      case Some(lastVersion) => new Random(lastVersion.hashCode()).nextBytes(version)
+      case None => Random.nextBytes(version)
+    }
+
+    new ByteArrayWrapper(version)
+  }
+
+  def update(epoch: ConsensusEpochNumber, boxMerklePathInfoSeq: Seq[BoxMerklePathInfo]): Try[ForgingBoxesMerklePathStorage] = Try {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: common behavior

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    this
+  }
+
+  def updateVersion(version: ByteArrayWrapper): Try[ForgingBoxesMerklePathStorage] = Try {
+    storage.update(version, new JArrayList[Pair[ByteArrayWrapper, ByteArrayWrapper]](), new JArrayList[ByteArrayWrapper]())
+    this
+  }
+
+  def getMerklePathsForEpoch(epoch: ConsensusEpochNumber): Option[Seq[BoxMerklePathInfo]] = {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: case, when no epoch info at all. Same answer as in State

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    storage.get(epochKey(epoch)) match {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: `storage.get()` returns `java.Optional`
to be discussed

File: sdk/src/main/scala/com/horizen/utils/BoxMerklePathInfo.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
+package com.horizen.utils
+
+import java.util
+
+import scorex.core.NodeViewModifier
+import scorex.core.serialization.{BytesSerializable, ScorexSerializer}
+import scorex.util.serialization.{Reader, Writer}
+
+
+case class BoxMerklePathInfo(boxId: Array[Byte], merklePath: MerklePath) extends BytesSerializable {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/utils/BoxMerklePathInfo.scala]: any reason to keep it as `ByteArrayWrapper`?

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainWallet.scala]:
````
walletBoxStorage.rollback(new ByteArrayWrapper(version)) match {
      case Failure(e) => throw e
    }
````

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: should be a part of discussion and task related to Storage interface changes

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: ok. as for State comments

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainWallet.scala]: It should be:
```
walletBoxStorage.rollback(new ByteArrayWrapper(version)) match {
      case Failure(e) => throw e
      case Success(_) =>
    }
```
but looks so ugly than
```
walletBoxStorage.rollback(new ByteArrayWrapper(version)).get
```
with the same sense.
I will leave it as is, but Storage interface should be discussed.

File: sdk/src/main/scala/com/horizen/SidechainWallet.scala [hunk 10]
```diff
@@ -41,8 +36,12 @@ trait Wallet[S <: Secret, P <: Proposition, TX <: Transaction, PMOD <: scorex.co
   def publicKeys(): Set[P]
 }
 
-class SidechainWallet private[horizen] (seed: Array[Byte], walletBoxStorage: SidechainWalletBoxStorage, secretStorage: SidechainSecretStorage,
-                                        walletTransactionStorage: SidechainWalletTransactionStorage, applicationWallet: ApplicationWallet)
+class SidechainWallet private[horizen] (seed: Array[Byte],
```
Comment (oaleksandrov-ip) [inline sdk/src/main/scala/com/horizen/SidechainWallet.scala]: we need information about spent up to two epochs ago ForgerBoxes

File: sdk/src/test/scala/com/horizen/SidechainWalletTest.scala [hunk 11]
```diff
@@ -627,4 +684,113 @@ class SidechainWalletTest
...
+
+
+    // Test 3: fifth epoch info - should request for the third epoch from storage
+    val fifthEpoch = ConsensusEpochNumber @@ 5
+    expectedEpochInfo = ConsensusEpochNumber @@ 3
+    assertTrue("No merkle path expected for unknown id.", sidechainWallet.getForgingBoxMerklePath(boxId, fifthEpoch).isEmpty)
+
+
+    // Test 4: get existent merkle path
+    assertTrue("Merkle path expected to be found.", sidechainWallet.getForgingBoxMerklePath(storedBoxId, fifthEpoch).isDefined)
```
Comment (oaleksandrov-ip) [inline sdk/src/test/scala/com/horizen/SidechainWalletTest.scala]: add tests about spent forger boxes

File: sdk/src/test/scala/com/horizen/SidechainWalletTest.scala [hunk 12]
```diff
@@ -627,4 +684,113 @@ class SidechainWalletTest
...
+
+        assertEquals("Different epoch number request expected.", epochNumber, epoch)
+        assertEquals("Different merkle path seq size expected.", 1, boxMerklePathInfoSeq.size)
+        assertArrayEquals("Different box id applied.", forgerBox1.id(), boxMerklePathInfoSeq.head.boxId)
+        assertArrayEquals("Wrong merkle path applied.", merkleTree.rootHash(), boxMerklePathInfoSeq.head.merklePath.apply(forgerBox1.id()))
+
+        Success(mockedForgingBoxesMerklePathStorage)
+      })
+
+    sidechainWallet.applyConsensusEpochInfo(epochInfo)
```
Comment (oaleksandrov-ip) [inline sdk/src/test/scala/com/horizen/SidechainWalletTest.scala]: check of stored consensus info information?

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+import scala.compat.java8.OptionConverters._
+import scala.util.{Failure, Random, Success, Try}
+import java.util.{ArrayList => JArrayList}
+
+
+class ForgingBoxesMerklePathStorage(storage: Storage) extends SidechainTypes with ScorexLogging
+{
+  require(storage != null, "Storage must be NOT NULL.")
+
+  private[horizen] def epochKey(epoch: ConsensusEpochNumber): ByteArrayWrapper = new ByteArrayWrapper(Blake2b256(s"epoch$epoch"))
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: for key or for related data?

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+          case Success(boxMerklePathsInfo) => Some(boxMerklePathsInfo.asScala)
+          case Failure(exception) =>
+            log.error("Error while box merkle paths info parsing.", exception)
+            None
+        }
+      case _ => None
+    }
+  }
+
+  def lastVersionId: Option[ByteArrayWrapper] = {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: agree

File: sdk/src/test/scala/com/horizen/SidechainWalletTest.scala (hunk 11 above)
Comment (i-Alex) [reply sdk/src/test/scala/com/horizen/SidechainWalletTest.scala]: ok.

File: sdk/src/test/scala/com/horizen/SidechainWalletTest.scala (hunk 12 above)
Comment (i-Alex) [reply sdk/src/test/scala/com/horizen/SidechainWalletTest.scala]: ok

Comment (i-Alex) [reply sdk/src/test/scala/com/horizen/SidechainWalletTest.scala]: found. it was done in testGetForgingBoxMerklePath. Check of what was applied to storage was done in line 779.

File: sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala [hunk 15]
```diff
@@ -55,10 +55,10 @@ class ForgingBoxesMerklePathStorage(storage: Storage) extends SidechainTypes wit
     this
   }
 
-  def getMerklePathsForEpoch(epoch: ConsensusEpochNumber): Option[Seq[BoxMerklePathInfo]] = {
+  def getInfoForEpoch(epoch: ConsensusEpochNumber): Option[Seq[ForgerBoxMerklePathInfo]] = {
```
Comment (oaleksandrov-ip) [inline sdk/src/main/scala/com/horizen/storage/ForgingBoxesMerklePathStorage.scala]: `getInfoForEpoch` -> `getForgerBoxMerklePathInfoForEpoch` ?

File: sdk/src/main/scala/com/horizen/utils/ForgerBoxMerklePathInfo.scala [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  override type M = ForgerBoxMerklePathInfo
+
+  override def serializer: ScorexSerializer[ForgerBoxMerklePathInfo] = ForgerBoxMerklePathInfoSerializer
+
+  override def hashCode: Int = forgerBox.hashCode()
+
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case info: ForgerBoxMerklePathInfo =>
+        forgerBox.equals(info.forgerBox) && merklePath.bytes().sameElements(info.merklePath.bytes())
```
Comment (oaleksandrov-ip) [inline sdk/src/main/scala/com/horizen/utils/ForgerBoxMerklePathInfo.scala]: btw just adding hash/equals to MerklePath is better idea

File: sdk/src/main/java/com/horizen/utils/MerklePath.java [hunk 17]
```diff
@@ -96,4 +91,34 @@ else if (size == 0)
...
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null)
+            return false;
+        if (!(this.getClass().equals(obj.getClass())))
+            return false;
+        if (obj == this)
```
Comment (oaleksandrov-ip) [inline sdk/src/main/java/com/horizen/utils/MerklePath.java]: better as a first or second check

**Pull request merged**